src/query/storage/converter.go

```diff
@@ -81,19 +88,104 @@ func PromTimeSeriesToSeriesAttributes(series prompb.TimeSeries) (ts.SeriesAttrib
 	default:
 		return ts.SeriesAttributes{}, fmt.Errorf("invalid source type %v", series.Source)
 	}
+
 	switch series.Type {
-	case prompb.Type_COUNTER:
-		metricType = ts.MetricTypeCounter
-	case prompb.Type_GAUGE:
-		metricType = ts.MetricTypeGauge
-	case prompb.Type_TIMER:
-		metricType = ts.MetricTypeTimer
+
+	case prompb.MetricType_UNKNOWN:
+		promMetricType = ts.PromMetricTypeUnknown
+
+	case prompb.MetricType_COUNTER:
+		promMetricType = ts.PromMetricTypeCounter
+		handleValueResets = true
+
+	case prompb.MetricType_GAUGE:
+		promMetricType = ts.PromMetricTypeGauge
+
+	case prompb.MetricType_HISTOGRAM:
+		promMetricType = ts.PromMetricTypeHistogram
+		handleValueResets = true
+
+	case prompb.MetricType_GAUGE_HISTOGRAM:
+		promMetricType = ts.PromMetricTypeGaugeHistogram
+		name := metricNameFromLabels(series.Labels)
+		handleValueResets = bytes.HasSuffix(name, promDefaultCountSuffix)
+
+	case prompb.MetricType_SUMMARY:
+		promMetricType = ts.PromMetricTypeSummary
+		handleValueResets = true
+
+	case prompb.MetricType_INFO:
+		promMetricType = ts.PromMetricTypeInfo
+
+	case prompb.MetricType_STATESET:
+		promMetricType = ts.PromMetricTypeStateSet
+
+	default:
+		return ts.SeriesAttributes{}, fmt.Errorf("invalid Prometheus metric type %v", series.Type)
+	}
+
+	switch series.M3Type {
+	case prompb.M3Type_M3_COUNTER:
+		m3MetricType = ts.M3MetricTypeCounter
+		if promMetricType == ts.PromMetricTypeUnknown && series.Source == prompb.Source_GRAPHITE {
+			promMetricType = ts.PromMetricTypeCounter
+		}
+	case prompb.M3Type_M3_GAUGE:
+		m3MetricType = ts.M3MetricTypeGauge
+		if promMetricType == ts.PromMetricTypeUnknown && series.Source == prompb.Source_GRAPHITE {
+			promMetricType = ts.PromMetricTypeGauge
+		}
+	case prompb.M3Type_M3_TIMER:
+		m3MetricType = ts.M3MetricTypeTimer
 	default:
-		return ts.SeriesAttributes{}, fmt.Errorf("invalid metric type %v", series.Type)
+		return ts.SeriesAttributes{}, fmt.Errorf("invalid M3 metric type %v", series.M3Type)
 	}
+
 	return ts.SeriesAttributes{
-		Type:   metricType,
-		Source: sourceType,
+		M3Type:            m3MetricType,
+		PromType:          promMetricType,
+		Source:            sourceType,
+		HandleValueResets: handleValueResets,
+	}, nil
+}
+
+// SeriesAttributesToAnnotationPayload converts ts.SeriesAttributes into an annotation.Payload.
```

Comment:
This only converts the prom type into the payload, do we want to add the m3 type explicitly too, or is the idea it just gets implicitly calculated on the translation from annotation payload to attributes? I'd prefer if we did it explicitly but either way I don't believe this is currently happening